### DISCUSSION TITLE
Sending received streaming response to another destination as streaming response without accumulating it

### DISCUSSION THREAD

**Discussion by renganathan-m** [fastapi/fastapi#12947] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
I came across a similar question; but the answer is not very clear/definitive.
```

### Description

Here is my requirement:

I run a web service in FastAPI (in machine-a).
In the above, I need to contact another external web service (running in machine-b) for one of my routes. 
The response that I receive from the external service, is a streaming one.

Now I need to pass the streaming response (received from the external one) to my client (say: browser) as streaming one.
And I want to send the received response (from the external service) to my client (browser) without accumulating it in machine-a.

Could someone help on this using FastAPI pls.

### Operating System

Windows

### FastAPI Version

0.110.1

### Pydantic Version

2.7.0

### Python Version

3.11.1

**Comment by luzzodev** [comment, marked as answer]
Hi, you can start from this sample

```python
async def stream_with_cleanup_gen(url: str):
    # Create an async client
    async with httpx.AsyncClient() as client:
        # Stream the response from the external URL
        async with client.stream("GET", url) as upstream_response:
            # Yield the response first if you need to access to response fields
            yield upstream_response

            # Stream the response data chunk by chunk
            async for chunk in upstream_response.aiter_text():
                yield chunk  # Yield each chunk


@router.get("/test")
async def get_streaming():
    url = "https://www.example.com"  # The URL to stream content from

    # Get the first value from the generator (which will be httpx.Response)
    response_generator = stream_with_cleanup_gen(url)
    response: httpx.Response = await anext(response_generator)  # Equivalent to next() in async generator

    # Set headers for the response
    response_headers = {}
    excluded_headers = {"content-length", "transfer-encoding", "connection"}
    for header, value in response.headers.items():
        if header.lower() not in excluded_headers:
            response_headers[header] = value

    return StreamingResponse(
        response_generator,
        status_code=response.status_code,
        media_type="text/html"
    )
```

**Comment by renganathan-m** [comment]
Thanks @luzzodev , I will start with this.

**Reply by luzzodev** [reply]
I'll add just a bit of context.

In generator function the yield of httpx.Response is usefull only if you need to happend other information to the response for example headers or status code as showed.

In generator function you should customize the iterator based on your usecase:

- aiter_bytes
- aiter_text
- aiter_lines
- aiter_raw



```
# Stream the response data chunk by chunk
async for chunk in upstream_response.aiter_text():  # customize here iterator.
    yield chunk  # Yield each chunk
```

You should of course handle all exceptions that can occur in the request.

The client creation is handled inside the generator and not outside in the endpoint because otherwise the context manager will exit and is not possible to iter over the response chunks.

```
    # Create an async client
    async with httpx.AsyncClient() as client:
        # Stream the response from the external URL
        async with client.stream("GET", url) as upstream_response:
            # Yield the response first if you need to access to response fields
            yield upstream_response
```


You should adjust the media_type according to the real type.
Do not use any middleware that require the full body or the response will be cached all in memory.

This is the snippet to test the endpoint. I took it from [here](https://github.com/fastapi/fastapi/discussions/10701).
```

import httpx
import time

t0 = time.time()
response = httpx.get('http://localhost:5000/test')
for chunk in response.iter_lines():
    print(f"Chunk (time {int((time.time()-t0)*1000)}ms): {chunk}")
```
